### PR TITLE
Dev/parser

### PR DESCRIPTION
**Pull request by haukurb**
Parser implementation

### CODE REVIEW

File: src/greynirseq/nicenlp/examples/constituency_parsing/predict_file.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    parser.add_argument("--data", type=str, required=True)
+    parser.add_argument("--input-path", type=str, required=False)
+    parser.add_argument("--checkpoint", type=str, required=True)
+    parser.add_argument("--max-sentences", type=int, required=False, default=32)
+    parser.add_argument("--output-path", type=str, required=False, default="-")
+
+    args = parser.parse_args()
+
+    assert args.max_sentences > 0
+    if args.output_path and args.output_path == "-":
```
Comment (peturorri) [inline src/greynirseq/nicenlp/examples/constituency_parsing/predict_file.py:58]: Er ekki fyrra tékkið óþarft? Það er default gildi.

File: src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        return [NonterminalNode(simple_label, children=new_children)]
+
+    # tree.pretty_print()
+    result = _inner(tree)
+    assert len(result) == 1
+    # tree.pretty_print()
+    # breakpoint()
+    tree = result[0]
+
+    # if any(t.text == "júní" for t in tree.leaves):
```
Comment (peturorri) [inline src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py:36]: Má þetta ekki fara? Og restin af þessu debug dóti?

File: src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                tree = simplify_node_tree(tree)
+                out_fh.write(tree.as_nltk_tree().pformat(margin=2 ** 100).strip())
+                out_fh.write("\n")
+
+
+if __name__ == "__main__":
+    import argparse
+
+    try:
+        import argcomplete  # noqa: F401
```
Comment (peturorri) [inline src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py:80]: Gerir þetta eitthvað ef það er aldrei notað?

File: src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    except ImportError as e:  # noqa: F841
```
Comment (peturorri) [inline src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py:81]: Ég fíla ekki að nota noqa á athugasemdir sem eru réttar!

File: src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        "--gold-dir",
+        type=str,
+        required=False,
+        default="psd",
+    )
+
+    args = parser.parse_args()
+    input_dir = Path(args.input)
+    output_dir = Path(args.output)
+    simplified_dir = output_dir
```
Comment (peturorri) [inline src/greynirseq/nicenlp/examples/constituency_parsing/score_parser.py:117]: Þetta lítur út eins og leifar af kóða sem er horfinn.

**Review by peturorri — COMMENTED** [review]
Þetta er rosa mikið að lesa í einu :/
Ég er ekki nógu mikið inni í þessu parsing dóti til að geta gefið gáfulegt feedback á útfærsluna sjálfa.

File: src/greynirseq/nicenlp/examples/constituency_parsing/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+WORKDIR /github/fairseq
+RUN git submodule sync \
+    && git submodule update --init --recursive \
+    && git fetch \
+    && git checkout 5a75b079bf8911a327940c28794608e003a9fa52 \
+    && python setup.py build_ext --inplace \
+    && python setup.py install
+
+WORKDIR /github
+RUN git clone --single-branch --branch parser-container https://github.com/mideind/greynirseq greynirseq
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/Dockerfile:34]: ætti að vera til núna

File: src/greynirseq/nicenlp/examples/constituency_parsing/pretrain.sh [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+SEED=1234
+TOTAL_NUM_UPDATES=200000 #
+WARMUP_UPDATES=1000     #  percent of the number of updates
+LR=5e-05                # Peak LR for polynomial LR scheduler.
+MAX_SENTENCES=50         # Batch size.
+ACCUM=1
+
+# Pretrained model
+RESTORE_PATH=/data/models/icebert/icebert-base-igc-ic3-rmhvocab-IceBERT/checkpoint38.pt
+RESTORE_PATH=/data/models/icebert/icebert-first-rmh_different_vocab/icebert-base-36k/model.pt
```
Comment (peturorri) [inline src/greynirseq/nicenlp/examples/constituency_parsing/pretrain.sh:15]: tvítekið

File: src/greynirseq/nicenlp/examples/constituency_parsing/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
+# Instructions for Miðeind neural parser (experimental)
+
+The docker container can be also be located on dockerhub (`docker pull mideind/neural-parser`). To load the saved image into docker use:
+
+```
+docker load -i neural-parser.tar
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/README.md:6]: af CLARIN

File: src/greynirseq/nicenlp/examples/constituency_parsing/predict_file.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                output_stream.write(pred_tree.as_nltk_tree().pformat(margin=2 ** 100).strip())
+                output_stream.write("\n")
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser("Parse a text file using a given parser checkpoint")
+    parser.add_argument("--data", type=str, required=True)
+    parser.add_argument("--input-path", type=str, required=False)
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/predict_file.py:56]: Júbb

File: src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  # Obtained from: https://stackoverflow.com/questions/60266215/shuffle-output-of-find-with-fixed-seed?noredirect=1&lq=1
+  openssl enc -aes-256-ctr -pass pass:"$1" -nosalt </dev/zero 2>/dev/null
+}
+
+set -ex
+
+### data is kept in many separate .gld files, we need to merge them
+rm -f "${DATA_VALID_GOLD_FILESNAMES}"
+rm -f "${DATA_FINETUNE}/text.txt" "${DATA_FINETUNE}/term.txt" "${DATA_FINETUNE}/nonterm.txt"
+if [ ! -f "${DATA_VALID_GOLD_FILESNAMES}" ] ; then
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh:44]: breytt

File: src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+## Prepare gold validation set
+while read -r FILEPATH ; do
+    python "${EXPORTER}" export "${FILEPATH}" \
+        "${DATA_VALID}/text.txt" "${DATA_VALID}/term.txt" "${DATA_VALID}/nonterm.txt" \
+        --binarize-trees --append --ignore-errors --error-log="${WDIR}/data/error_trees.valid.psd" --append-errors
+done < "${DATA_VALID_GOLD_FILESNAMES}"
+
+# Prepare test set (using the gold testset)
+rm -f "${DATA_TEST}/text" "${DATA_TEST}/nonterm" "${DATA_TEST}/term"
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh:78]: breytt

File: src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+done
+
+# Prepare silver set (selected data from cfg parser)
+python "${EXPORTER}" export "${DATA_SILVER_PSD}" \
+    "${DATA_SILVER}/text.txt" "${DATA_SILVER}/term.txt" "${DATA_SILVER}/nonterm.txt" \
+    --binarize-trees --ignore-errors --error-log="${WDIR}/data/error_trees.silver.psd" --limit 1000000
+
+# Prepare copper set (rest of data from cfg parser)
+mkdir -p "${DATA_BASE}/data/copper"
+rm -f "${WDIR}/data/copper/nonterm.txt" "${WDIR}/data/copper/term.txt" > "${WDIR}/data/copper/text.txt"
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh:92]: breytt

File: src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+fi
+
+##############################################################################################
+###############    Prepare text, nonterm and term data from .psd fiels    ####################
+##############################################################################################
+
+### Prepare small dataset for debug purposes
+python "${EXPORTER}" export "${DATA_DEBUG_PSD}" \
+    "${DATA_DEBUG}/text.txt" "${DATA_DEBUG}/term.txt" "${DATA_DEBUG}/nonterm.txt" \
+    --binarize-trees --ignore-errors --error-log="${WDIR}/data/error_trees_gold.psd" --append-errors
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/prepare_data.sh:61]: það varðveitist milli keyrslna, það er viljandi

File: src/greynirseq/nicenlp/examples/constituency_parsing/pretrain.sh (hunk 7 above)
Comment (haukurb) [reply src/greynirseq/nicenlp/examples/constituency_parsing/pretrain.sh:15]: breytt

File: src/greynirseq/nicenlp/data/datasets.py [hunk 14]
```diff
@@ -4,15 +4,17 @@
 
 from collections import OrderedDict
 from functools import lru_cache
-from typing import Callable, Dict, List, Tuple
+from typing import Callable, Dict, List, Tuple  # noqa: F401
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/data/datasets.py:7]: Noqa?

Comment (peturorri) [reply src/greynirseq/nicenlp/data/datasets.py:7]: þetta slekkur á lint viðvörunum, í þessu dæmi "unused import" frá flake8

**Comment by HaukurPall** [comment]
Já, einmitt. Mér fannst bara skrítið að setja þetta á typing import.

**Comment by HaukurPall** [comment]
Allt komið frá mér. Buildar fínt hjá mér.

**Comment by HaukurPall** [comment]
@vesteinn @haukurb hverjum þarf ég að múta til þess að fá þetta inn?

**Review by vesteinn — APPROVED** [review]
